docs/exts/exampleinclude.py

```diff
@@ -0,0 +1,189 @@
+from os import path
+
+from docutils import nodes
+from docutils.parsers.rst import directives
+from six import text_type
+from sphinx import addnodes
+from sphinx.directives.code import LiteralIncludeReader
+from sphinx.locale import _
+from sphinx.pycode import ModuleAnalyzer
+from sphinx.util import logging
+from sphinx.util import parselinenos
+from sphinx.util.docutils import SphinxDirective
+from sphinx.util.nodes import set_source_info
+
+logger = logging.getLogger(__name__)
+
+
+class example_header(nodes.reference, nodes.FixedTextElement):
+    pass
+
+
+class ExampleInclude(SphinxDirective):
```

Comment:
Can we do this via subclassing LiteralInclude so we don't have to have so much code here?

Reply:
I know that we have a lot of code, but it is not possible to extend this class. The logic of this directive is contained in one method. There is no method that we can extend to add new expected  behavior.

Reply:
If we subclass we at least don't have to set the `option_spec` and other class properties.

Reply:
Which version did you copy from? There are odd/subtle differences from the version linked to

Reply:
Unless I'm missing something (I've only had a _quick_ look) I think we should be able to do this as something like:

```python
    def run(self):
        _, filename = self.env.relfn2path(self.arguments[0])
        retnode, = super().run()
        container_node = nodes.container("", literal_block=True, classes=["example-block-wrapper"])
        container_node += example_header(filename=filename)
        container_node += retnode
        retnode = container_node
        return [retnode]
```

?

Reply:
I based on:
https://github.com/sphinx-doc/sphinx/blob/v1.8.3/sphinx/directives/code.py
I've automatically formatted the code to be compatible with my company's requirements. We use black. Changes from my team are undergoing internal review. During it, we had a lot of discussion about style, which forced us to introduce automatic formatting of the code.

I made changes in the `option_spec` field. If I extend this class then I will have to make identical modifications programmatically. It is worth noting that the value of this field is closely related to the implementation. If the value of the field changes, then the implementation must also be changed.

Reply:
My changes affect these lines:
https://github.com/sphinx-doc/sphinx/blob/v1.8.3/sphinx/directives/code.py#L456-L462

Reply:
(I figured there was some auto-formatting going on)

A link to that would more helpful than master :)

The only change in option_spec is the removal of `caption` right? Basically there is a lot of code here that we haven't really changed, and it's hard to work out what is special and what isn't.

Reply:
I will prepare a detailed description of the changes that I made in the code of this class

Reply:
Can you wait with review? I did not think that this change might require such a discussion

Reply:
I might be over thinking it - still, a comment in the source  would help future Us.

Reply:
Diff changes: https://pastebin.com/W4eHex7C
Line 1-5: I changed the name of the class. No comemnts
Line 6-15: I updated description of class. No comments
Line 16-17: I remove caption. Now, we use a filename as a caption allways.
Line 18-19 At the time of writing PR, there were no validation types in the project.
Line 20-25: There are two changes here.
a) I dropped support for translating the message. We do not support translations in the project.
b) I pass a arguments to logger. Reason available: https://github.com/apache/airflow/pull/4804
Line 26-38. 
a) I dropped caption support. It was conflicting with the new header.
b) I added a new header.  This uses the file name as the header content.

> Unless I'm missing something (I've only had a quick look) I think we should be able to do this as something like:

I did not find the official API to make a distinction.  This is just one class and I decided it would be better to copy it. I've done research on how other people implement extensions for this directive and other people also copy the full class code
https://github.com/FabriceSalvaire/sphinx-getthecode/blob/master/sphinxcontrib/getthecode.py#L27-L32